dynamodb/data.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+// ProcessData performs data conversion for DynamoDB database. For each table,
+// we extract data using Scan requests, convert the data to Spanner data (based
+// on the source and Spanner schemas), and write it to Spanner. If we can't
+// get/process data for a table, we skip that table and process the remaining
+// tables.
+func ProcessData(conv *internal.Conv, client dynamoClient, consistentRead bool) error {
+	for srcTable, srcSchema := range conv.SrcSchema {
+		spTable, err1 := internal.GetSpannerTable(conv, srcTable)
+		spCols, err2 := internal.GetSpannerCols(conv, srcTable, srcSchema.ColNames)
+		spSchema, ok := conv.SpSchema[spTable]
+		if err1 != nil || err2 != nil || !ok {
+			conv.Stats.BadRows[srcTable] += conv.Stats.Rows[srcTable]
+			conv.Unexpected(fmt.Sprintf("Can't get cols and schemas for table %s: err1=%s, err2=%s, ok=%t",
+				srcTable, err1, err2, ok))
+			continue
+		}
+
+		var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+
+		for {
+			// Build the query input parameters
+			params := &dynamodb.ScanInput{
+				TableName:      aws.String(srcTable),
+				ConsistentRead: &consistentRead,
+			}
+			if lastEvaluatedKey != nil {
+				params.ExclusiveStartKey = lastEvaluatedKey
+			}
+
+			// Make the DynamoDB Query API call
+			result, err := client.Scan(params)
+			if err != nil {
+				return fmt.Errorf("failed to make Query API call for table %v: %v", srcTable, err)
+			}
+
+			// Iterate the items returned
+			for _, attrsMap := range result.Items {
+				var cvtVals []interface{}
+				for i, srcColName := range srcSchema.ColNames {
+					var cvtVal interface{}
+					if attrsMap[srcColName] == nil {
+						cvtVal = nil
+					} else {
+						// Convert data to the target type
+						cvtVal, err = cvtColValue(attrsMap[srcColName], srcSchema.ColDefs[srcColName].Type.Name, spSchema.ColDefs[spCols[i]].T.Name)
+						if err != nil {
+							return fmt.Errorf("failed to convert column: %v to %v", attrsMap[srcColName], spSchema.ColDefs[spCols[i]])
+						}
+					}
+					cvtVals = append(cvtVals, cvtVal)
+				}
+
+				conv.WriteRow(srcTable, spTable, spCols, cvtVals)
+			}
+
+			if result.LastEvaluatedKey == nil {
+				break
+			}
+			// If there are more rows, then continue.
+			lastEvaluatedKey = result.LastEvaluatedKey
+		}
+	}
+
+	return nil
+}
+
+func cvtColValue(attrVal *dynamodb.AttributeValue, srcType string, spType string) (interface{}, error) {
+	switch spType {
+	case ddl.Bool:
+		switch srcType {
+		case typeBool:
+			return *attrVal.BOOL, nil
+		}
+	case ddl.Bytes:
+		switch srcType {
+		case typeBinary:
+			return attrVal.B, nil
+		case typeBinarySet:
+			return attrVal.BS, nil
+		}
+	case ddl.String:
+		switch srcType {
+		case typeMap:
+			b, err := json.Marshal(attrVal)
+			if err != nil {
+				return nil, fmt.Errorf("failed to encode a map object: %v to a json string", attrVal.GoString())
```

Comment:
From a user point of view, we're doing data conversion in this code, so the use of "encoding" here might be confusing.
Maybe "convert"?
Same comment for the three other error messages.